enhancements/cluster-runtime-constraints.md

```diff
@@ -0,0 +1,138 @@
+
+### Implementation Details
+
+#### Cluster Runtime Constraint Configmap
+
+Cluster admins can specified cluster runtime constraints in a configmap (named `olm-runtime-constraints`) in `olm` namespace using this format:
+
+```yaml=
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: olm-runtime-constraints
+  namespace: olm
+data:
+    properties: <list-of-expression-constraints-in-json>
+```
+
+Each constraint is specified using the following syntax:
+
+```yaml=
+{
+  "type": "olm.constraint",
+  "value": {
+    "evaluator": {
+        "id": "cel",
+    }
+    "source": 'properties.exists(p, p.type == "certified")',
+    "action": {
+        "id": "require"
+    }
+  }
+}
+```
+
+The constraint `type` is `olm.constraint` and the `value` is information associated with constraint. The `value` struct has 3 fields: `evaluator`, `source` and `action`.
+
+The `evaluator` is a struct with `id` field to represent the languague library that will be used to evaluate the expression. At the moment, only `cel-go` library is supported using `cel` as the identifier. More fields are potentially added to expand the supported expression languagues/libraries in the future.
+
+The `source` field is the string of expression that will be evaluated during resolution. Currently, only `cel-go` expressions are supported.
+
+Finally, the `action` field is a struct with the `id` field to indicate the resolution action that this constraint will behave. For example, for `require` action, there must be at least one candidate that satisfies this constraint. At the moment, `require` and `conflict` actions are supported. More fields may be added into `action` struct to support more resolution actions in the future.
+
+#### Runtime Dependencies in Bundle
+
+The operator authors specify the cluster runtime requirements in `dependencies.yaml` using the following format:
+
+```yaml=
+dependencies:
+  -
+    type: olm.constraint
+    value:
+      action:
+        id: require
+      evaluator:
+        id: cel
+      source: "properties.exists(p, p.type == \"certified\")"
+```
+
+#### Runtime Constraint Resolution
+
+The constraints that are specified in `olm-runtime-constraints` Configmap are converted into properties of a global existing virtual node in the cluster. All operators/candidates must satisfy those properties in order to be installed.
+
+If constraints are added to bundles, they will be evaluated and must be satisfied similiarly to `olm.gvk.required` and `olm.package.required` propertoes.
+
+### Risks and Mitigations
+
+### Prototype
+
+TBD
+
+## Drawbacks
+
+The [cel-go](https://github.com/google/cel-go) library is rather new and still under development. Even though cel-go has been used in other opensource projects such as Tekton, it has not yet reached stable release (v1). Breaking changes can be introduced to this library and potentially OLM behavior in the future.
```

Comment:
It would be useful to add some explanation of why this isn't a deal-breaker.

Reply:
This particular library has been used in some other well-known projects such as Istio, caddy, grpc and etc. Those are stable projects and it hasn't been a problem for them. Frankly, OLM itself is still an under-development project as it hasn't reached v1 release yet either. I'm not saying that we don't care about breaking changes but we can at least pin this project to a version that we are comfortable with. I expect we will potentially support other libraries/languages in the future as I mentioned in the `evaluator` spec. We can introduce a more mature library later and deprecate cel-go if it reaches the point that we can no longer support it in the long run.

Reply:
Thanks. I suggest adding that to the proposal. Since OLM is part of openshift, even if its APIs are not yet v1, backward compatibility still needs to be handled with care.